lxd/devlxd_storage.go

```diff
@@ -46,3 +47,83 @@ func devLXDStoragePoolGetHandler(d *Daemon, r *http.Request) response.Response {
 
 	return response.DevLXDResponseETag(http.StatusOK, respPool, "json", etag)
 }
+
+var devLXDStoragePoolVolumesEndpoint = devLXDAPIEndpoint{
+	Path: "storage-pools/{poolName}/volumes",
+	Get:  devLXDAPIEndpointAction{Handler: devLXDStoragePoolVolumesGetHandler, AccessHandler: allowDevLXDAuthenticated},
+}
+
+var devLXDStoragePoolVolumesTypeEndpoint = devLXDAPIEndpoint{
+	Path: "storage-pools/{poolName}/volumes/{type}",
+	Get:  devLXDAPIEndpointAction{Handler: devLXDStoragePoolVolumesGetHandler, AccessHandler: allowDevLXDAuthenticated},
+}
+
+func devLXDStoragePoolVolumesGetHandler(d *Daemon, r *http.Request) response.Response {
+	inst, err := getInstanceFromContextAndCheckSecurityFlags(r.Context(), devLXDSecurityKey, devLXDSecurityMgmtVolumesKey)
+	if err != nil {
+		return response.DevLXDErrorResponse(err)
+	}
+
+	poolName := mux.Vars(r)["poolName"]
+	volType := mux.Vars(r)["type"]
+	projectName := inst.Project().Name
+
+	// Get identity from the request context.
+	identity, err := getDevLXDIdentity(r.Context())
+	if identity == nil {
+		return response.DevLXDErrorResponse(err)
+	}
+
+	// Reject non-recursive requests.
+	if !util.IsRecursionRequest(r) {
+		return response.DevLXDErrorResponse(api.NewStatusError(http.StatusNotImplemented, "Only recursive requests are currently supported"))
+	}
+
+	// Reject non-custom volume types, if the type is specified.
+	if volType != "" && volType != "custom" {
+		return response.DevLXDErrorResponse(api.NewStatusError(http.StatusBadRequest, "Only custom storage volumes can be retrieved"))
+	}
+
+	// Get storage volumes.
+	vols := []api.StorageVolume{}
+
+	url := api.NewURL().Path("1.0", "storage-pools", poolName, "volumes", volType).Project(projectName).WithQuery("recursion", "1")
+	target := r.URL.Query().Get("target")
+	if target != "" {
+		url = url.WithQuery("target", target)
+	}
+
+	// Ensure only custom volumes are returned, if the volume type is not provided.
+	if volType == "" {
+		url = url.WithQuery("filter", "type eq custom")
+	}
+
+	req, err := NewRequestWithContext(r.Context(), http.MethodGet, url.String(), nil, "")
+	if err != nil {
+		return response.DevLXDErrorResponse(err)
+	}
+
+	resp := storagePoolVolumesGet(d, req)
+	_, err = RenderToStruct(req, resp, &vols)
+	if err != nil {
+		return response.DevLXDErrorResponse(err)
+	}
+
+	respVols := make([]api.DevLXDStorageVolume, 0, len(vols))
+	for _, vol := range vols {
+		if vol.Config["volatile.owner"] != identity.Identifier {
```

Comment:
This is easy to miss, lets break it out into a separate function and then add a comment explaining what is happening here.